## **1) Problem statement**
This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.
## **2) Data Collection**
Dataset Source - https://github.com/Akashram28/Cranklever_Datahack/blob/main/datahack_videos.csv
The data consists of 8 column and 1000 rows.
## **2.1 Import Data and Required Packages**
Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.


In [461]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Import the CSV Data as Pandas DataFrame

In [462]:
df = pd.read_csv('INvideos.csv')
dur = pd.read_csv('duration.csv')
df['duration']=dur['duration']

### Show Top 5 Records

In [463]:
df.head()

video_id trending_date  \
0  kzwfHumJyYc      17.14.11   
1  zUZ1z7FwLc8      17.14.11   
2  10L1hZ9qa58      17.14.11   
3  N1vE8iiEg64      17.14.11   
4  kJzGH0PVQHQ      17.14.11   

                                               title    channel_title  \
0  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...  Lokdhun Punjabi   
1  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...          HJ NEWS   
2  Stylish Star Allu Arjun @ ChaySam Wedding Rece...             TFPC   
3                     Eruma Saani | Tamil vs English      Eruma Saani   
4  why Samantha became EMOTIONAL @ Samantha naga ...       Filmylooks   

   category_id              publish_time  \
0            1  2017-11-12T12:20:39.000Z   
1           25  2017-11-13T05:43:56.000Z   
2           24  2017-11-12T15:48:08.000Z   
3           23  2017-11-12T07:08:48.000Z   
4           24  2017-11-13T01:14:16.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   
2  Stylish Star Allu Arjun @ ChaySam Wedding Rece...   473988   2011   
3  Eruma Saani|"Tamil Comedy Videos"|"Films"|"Mov...  1242680  70353   
4  Filmylooks|"latest news"|"telugu movies"|"telu...   464015    492   

   dislikes  comment_count                                  thumbnail_link  \
0       798            882  https://i.ytimg.com/vi/kzwfHumJyYc/default.jpg   
1       904              0  https://i.ytimg.com/vi/zUZ1z7FwLc8/default.jpg   
2       243            149  https://i.ytimg.com/vi/10L1hZ9qa58/default.jpg   
3      1624           2684  https://i.ytimg.com/vi/N1vE8iiEg64/default.jpg   
4       293             66  https://i.ytimg.com/vi/kJzGH0PVQHQ/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1               True             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description duration  
0  Presenting Sharry Mann latest Punjabi Song  Cu...    PT27S  
1  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...     PT2M  
2  Watch Stylish Star Allu Arjun @ ChaySam Weddin...   PT2M9S  
3  This video showcases the difference between pe...  PT4M36S  
4  why Samantha became EMOTIONAL @ Samantha naga ...   PT1M6S

### Shape of the dataset

In [464]:
df.shape

(37352, 17)

## **3. Data Checks to perform**
- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

## **3.1 Check Missing values**

In [465]:
df.isna().sum()

video_id                      0
trending_date                 0
title                         0
channel_title                 0
category_id                   0
publish_time                  0
tags                          0
views                         0
likes                         0
dislikes                      0
comment_count                 0
thumbnail_link                0
comments_disabled             0
ratings_disabled              0
video_error_or_removed        0
description                 561
duration                  16523
dtype: int64

There are some missing values in dataset

## 3.2 Check Duplicates

In [466]:
df.duplicated().sum()

1785

There are duplicate values in the dataset

In [467]:
df.drop(['trending_date','comments_disabled','ratings_disabled','video_error_or_removed','video_id'],axis=1)
df.head(2)

video_id trending_date  \
0  kzwfHumJyYc      17.14.11   
1  zUZ1z7FwLc8      17.14.11   

                                               title    channel_title  \
0  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...  Lokdhun Punjabi   
1  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...          HJ NEWS   

   category_id              publish_time  \
0            1  2017-11-12T12:20:39.000Z   
1           25  2017-11-13T05:43:56.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes  comment_count                                  thumbnail_link  \
0       798            882  https://i.ytimg.com/vi/kzwfHumJyYc/default.jpg   
1       904              0  https://i.ytimg.com/vi/zUZ1z7FwLc8/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1               True             False                   False   

                                         description duration  
0  Presenting Sharry Mann latest Punjabi Song  Cu...    PT27S  
1  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...     PT2M

## 3.3 Checking Data Types

In [468]:
df.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
duration                  object
dtype: object

There are 3 types of data present : object,int64,bool

## 3.4 Checking Unique Values

In [469]:
df.nunique()

video_id                  16307
trending_date               205
title                     16721
channel_title              1426
category_id                  17
publish_time              16339
tags                      12578
views                     32136
likes                     15529
dislikes                   5079
comment_count              6027
thumbnail_link            16523
comments_disabled             2
ratings_disabled              2
video_error_or_removed        2
description               13992
duration                   2203
dtype: int64

## 3.5 Checking stats of Data Set

In [470]:
df.describe()

category_id         views         likes      dislikes  comment_count
count  37352.000000  3.735200e+04  3.735200e+04  3.735200e+04    37352.00000
mean      21.576596  1.060478e+06  2.708272e+04  1.665082e+03     2676.99743
std        6.556593  3.184932e+06  9.714510e+04  1.607617e+04    14868.31713
min        1.000000  4.024000e+03  0.000000e+00  0.000000e+00        0.00000
25%       23.000000  1.239155e+05  8.640000e+02  1.080000e+02       81.00000
50%       24.000000  3.045860e+05  3.069000e+03  3.260000e+02      329.00000
75%       24.000000  7.992912e+05  1.377425e+04  1.019250e+03     1285.00000
max       43.000000  1.254322e+08  2.912710e+06  1.545017e+06   827755.00000

### **Insight**

- On average a trending video gets around 1 million views
- Maximum View a trending video got was 1.25 billion
- Trending Videos gets more likes than dislikes
- Trending videos gets around 2677 comments
 

## **4. Cleaning Data**

removing unnecessary features

In [471]:
df = df.drop(['trending_date','comments_disabled','ratings_disabled','video_error_or_removed','thumbnail_link','comment_count'],axis=1)
df.dropna(inplace=True)
df.head(2)

video_id                                              title  \
0  kzwfHumJyYc  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1  zUZ1z7FwLc8  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   

     channel_title  category_id              publish_time  \
0  Lokdhun Punjabi            1  2017-11-12T12:20:39.000Z   
1          HJ NEWS           25  2017-11-13T05:43:56.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes                                        description duration  
0       798  Presenting Sharry Mann latest Punjabi Song  Cu...    PT27S  
1       904  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...     PT2M

## Converting duration into appropriate format

In [472]:
from datetime import datetime, time
def change(l):
    m=0
    t=""
    s=0
    h=0
    l=l[2:]
    for i in l:
        if i=='D':
            t=""
            continue
        if i=='T':
            t=""
            continue
        if i=='H':
            h=int(t)
            t=""
            continue
        if i=='M':
            m=int(t)
            t=""
            continue
        if i=='S':
            s=int(t)
            t=""
        t+=i
    duration=time(h,m,s)
    return duration

df['duration']=df['duration'].apply(change)
df.head(2)

video_id                                              title  \
0  kzwfHumJyYc  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1  zUZ1z7FwLc8  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   

     channel_title  category_id              publish_time  \
0  Lokdhun Punjabi            1  2017-11-12T12:20:39.000Z   
1          HJ NEWS           25  2017-11-13T05:43:56.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes                                        description  duration  
0       798  Presenting Sharry Mann latest Punjabi Song  Cu...  00:00:27  
1       904  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...  00:02:00

In [473]:
def total_seconds(i):
    t = i.hour * 3600 + i.minute*60 + i.second
    return t
df['duration']=df['duration'].apply(total_seconds)
df.head(2)

video_id                                              title  \
0  kzwfHumJyYc  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1  zUZ1z7FwLc8  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   

     channel_title  category_id              publish_time  \
0  Lokdhun Punjabi            1  2017-11-12T12:20:39.000Z   
1          HJ NEWS           25  2017-11-13T05:43:56.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes                                        description  duration  
0       798  Presenting Sharry Mann latest Punjabi Song  Cu...        27  
1       904  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...       120

## **Converting category_id to Category title**

getting title for each id via web scraping


In [474]:
import requests
from bs4 import BeautifulSoup as bs

r=requests.get('https://mixedanalytics.com/blog/list-of-youtube-video-category-ids/')

soup=bs(r.content)

In [475]:
table=soup.find('table')
rows=table.findAll('td')
d={}
for i,v in enumerate(rows):
    if i%2==0:
        try:
            d[int(v.getText())]=rows[i+1].getText()
        except:
            continue
print(d)

{1: 'Film & Animation', 2: 'Autos & Vehicles', 10: 'Music', 15: 'Pets & Animals', 17: 'Sports', 18: 'Short Movies', 19: 'Travel & Events', 20: 'Gaming', 21: 'Videoblogging', 22: 'People & Blogs', 23: 'Comedy', 24: 'Entertainment', 25: 'News & Politics', 26: 'Howto & Style', 27: 'Education', 28: 'Science & Technology', 29: 'Nonprofits & Activism', 30: 'Movies', 31: 'Anime/Animation', 32: 'Action/Adventure', 33: 'Classics', 34: 'Comedy', 35: 'Documentary', 36: 'Drama', 37: 'Family', 38: 'Foreign', 39: 'Horror', 40: 'Sci-Fi/Fantasy', 41: 'Thriller', 42: 'Shorts', 43: 'Shows', 44: 'Trailers'}


In [476]:
def title(i):
    if i in d:
        return d[i]
df['category_id']=df['category_id'].apply(title)
df.rename(columns= {'category_id':'category'},inplace=True)
df.head(2)

video_id                                              title  \
0  kzwfHumJyYc  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1  zUZ1z7FwLc8  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   

     channel_title          category              publish_time  \
0  Lokdhun Punjabi  Film & Animation  2017-11-12T12:20:39.000Z   
1          HJ NEWS   News & Politics  2017-11-13T05:43:56.000Z   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes                                        description  duration  
0       798  Presenting Sharry Mann latest Punjabi Song  Cu...        27  
1       904  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...       120

## **Changing publish_time to categorical feature**

- 4:00-11:00 --> Morning
- 11:00-16:00 --> Afternoon
- 16:00-21:00 --> Evening
- 21:00-4:00 --> Night



In [477]:
def change1(i):
    if 4<=i.hour<11:
        return 'Morning'
    elif 11<=i.hour<16:
        return 'Afternoon'
    elif 16<=i.hour<22:
        return 'Evening'
    else:
        return 'Night'

df['publish_time']=df['publish_time'].apply(pd.to_datetime)
df['publish_time']=df['publish_time'].apply(change1)
df.head(2)

video_id                                              title  \
0  kzwfHumJyYc  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1  zUZ1z7FwLc8  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   

     channel_title          category publish_time  \
0  Lokdhun Punjabi  Film & Animation    Afternoon   
1          HJ NEWS   News & Politics      Morning   

                                                tags    views  likes  \
0  sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1  पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   

   dislikes                                        description  duration  
0       798  Presenting Sharry Mann latest Punjabi Song  Cu...        27  
1       904  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...       120

## **Exploring Data (Visualisation)**

In [478]:
import plotly.express as px
fig = px.histogram(df,x='publish_time',y='views',barmode='group')
fig.show()

In [479]:
fig = px.histogram(df,x='category',y='views',barmode='group')
fig.show()


In [480]:
df1 = df
category = ['Science & Technology','Howto & Style','Shows','Nonprofits & Activism','Autos & Vehicles','Pets & Animals','Movies']
ind = []
for i in category:
    ind.append(df1[df1['category']== i].index.tolist())
Ind = []
for i in ind:
   for _ in i:
        Ind.append(_)
df1.drop(Ind,inplace=True)


In [481]:
fig = px.histogram(df1,x='publish_time',y='views',color='category', barmode='group',width=1150,height=800)
fig.show()